File: webgui/src/app/modules/fuel-station/mappers/fuel-station.mapper.ts
import { FuelStation } from "../models/fuel-station.model";
import FuelStationStatus from "../models/fuel-station-status.enum";
import FuelPrice from "../models/fuel-price.model";
import { FuelTank } from "../models/fuel-tank.model";
import { FuelGradeMapper } from "../../common/fuel-grade.mapper";

export class FuelStationMapper {
  static fromJson(json: unknown): FuelStation {
    const typedJson = json as Record<string, unknown>
    const status = this.parseStatus(typedJson['status']);
    const fuelPrices = this.parseFuelPrices(typedJson['fuelPrices'] as Record<string,unknown>[]);
    const fuelTanks = this.parseFuelTanks(typedJson['fuelTanks'] as Record<string,unknown>[] );

    return new FuelStation(
      Number(typedJson['id']),
      typedJson['street'] as string,
      typedJson['buildingNumber'] as string,
      typedJson['city'] as string,
      typedJson['postalCode'] as string,
      typedJson['country'] as string,
      typedJson['address'] as string,
      fuelPrices,
      fuelTanks,
      typedJson['assignedManagersIds'] as number[],
      status
    );
  }

  private static parseStatus(status: unknown): FuelStationStatus {
    if (typeof status === 'string') {
      switch(status) {
      case "active":
        return FuelStationStatus.Active;
      case "deactivated":
        return FuelStationStatus.Deactivated
      default:
        throw new Error(`Unsupported status type: ${typeof status}`);

      }
    } 

    throw new Error(`Unsupported status type: ${typeof status}`);
  }

  private static parseFuelPrices(data: Record<string, unknown>[]): FuelPrice[] {
    return data.map(json => {
      const fuelGrade = FuelGradeMapper.map(json['fuelGrade']);
      return new FuelPrice(fuelGrade, Number(json['pricePerLiter']));
    });
  }

  private static parseFuelTanks(data: Record<string, unknown>[]): FuelTank[] {
    return data.map(json => {
      const fuelGrade = FuelGradeMapper.map(json['fuelGrade']);
      return new FuelTank(
        Number(json['id']),
        Number(json['currentVolume']),
        Number(json['maxCapacity']),
        fuelGrade,
        json['lastRefillDate'] ? new Date(json['lastRefillDate'] as string) : null
      );
    });
  }

}

File: fsms-backend-api/src/main/java/com/fuelstation/managmentapi/fuelstation/infrastructure/persistence/FuelStationMapper.java
package com.fuelstation.managmentapi.fuelstation.infrastructure.persistence;

import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

import com.fuelstation.managmentapi.country.domain.CountryCode;
import com.fuelstation.managmentapi.common.domain.CurrencyCode;
import com.fuelstation.managmentapi.fuelgrade.domain.FuelGrade;
import com.fuelstation.managmentapi.fuelstation.domain.models.*;
import com.fuelstation.managmentapi.fuelstation.infrastructure.persistence.entity.FuelPriceEmbeddable;
import com.fuelstation.managmentapi.fuelstation.infrastructure.persistence.entity.FuelStationAddressEmbeddable;
import com.fuelstation.managmentapi.fuelstation.infrastructure.persistence.entity.FuelStationEntity;
import com.fuelstation.managmentapi.fuelstation.infrastructure.persistence.entity.FuelTankEntity;
import org.springframework.stereotype.Component;

@Component
public class FuelStationMapper {

    public FuelStation toDomain(FuelStationEntity entity) {
        FuelStationAddressEmbeddable addressEmbeddable = entity.getAddress();
        FuelStationAddress address = new FuelStationAddress(
                addressEmbeddable.getStreet(),
                addressEmbeddable.getBuildingNumber(),
                addressEmbeddable.getCity(),
                addressEmbeddable.getPostalCode(),
                CountryCode.fromId(addressEmbeddable.getCountryId())
        );

        List<FuelTank> fuelTanks = entity.getFuelTanks().stream()
                .filter(tankEntity -> FuelTankStatus.fromId(tankEntity.getFuelTankStatusId()) == FuelTankStatus.ACTIVE)
                .map(tankEntity -> new FuelTank(
                        tankEntity.getFuelTankId(),
                        FuelGrade.fromId(tankEntity.getFuelGradeId()),
                        tankEntity.getCurrentVolume(),
                        tankEntity.getMaxCapacity(),
                        tankEntity.getLastRefillDate() != null ?
                                Optional.of(tankEntity.getLastRefillDate()) :
                                Optional.empty(),
                        FuelTankStatus.fromId(tankEntity.getFuelTankStatusId())
                ))
                .collect(Collectors.toList());

        List<FuelStationFuelPrice> fuelPrices = entity.getFuelPrices().stream()
                .map(priceEmbeddable -> new FuelStationFuelPrice(
                        FuelGrade.fromId(priceEmbeddable.getFuelGradeId()),
                        priceEmbeddable.getPricePerLiter(),
                        CurrencyCode.fromString(priceEmbeddable.getCurrency())
                ))
                .collect(Collectors.toList());

        List<Long> managerIds = entity.getAssignedManagers();

        return new FuelStation(
                entity.getFuelStationId(),
                address,
                fuelTanks,
                fuelPrices,
                managerIds,
                FuelStationStatus.fromId(entity.getFuelStationStatusId()),
                entity.getCreatedAt()
        );
    }

    public FuelStationEntity toEntity(FuelStation domain) {

        FuelStationAddress address = domain.getAddress();
        FuelStationAddressEmbeddable addressEmbeddable = new FuelStationAddressEmbeddable(
                address.street(),
                address.buildingNumber(),
                address.city(),
                address.postalCode(),
                address.country().getId()
        );

        FuelStationEntity entity = new FuelStationEntity();
        entity.setFuelStationId(domain.getFuelStationId());
        entity.setAddress(addressEmbeddable);
        entity.setFuelStationStatusId(domain.getStatus().getId());
        entity.setCreatedAt(domain.getCreatedAt());

        List<FuelPriceEmbeddable> fuelPriceEmbeddables = domain.getFuelPrices().stream()
                .map(price -> new FuelPriceEmbeddable(
                        price.fuelGrade().getId(),
                        price.pricePerLiter(),
                        price.currency().name()
                ))
                .collect(Collectors.toList());
        entity.setFuelPrices(fuelPriceEmbeddables);

        entity.setAssignedManagers(domain.getAssignedManagersIds());

        List<FuelTankEntity> fuelTankEntities = domain.getFuelTanks().stream()
                .map(tank -> new FuelTankEntity(
                        tank.getId(),
                        tank.getFuelGrade().getId(),
                        tank.getCurrentVolume(),
                        tank.getMaxCapacity(),
                        entity,
                        tank.getLastRefillDate().orElse(null),
                        tank.getStatus().getId()
                ))
                .collect(Collectors.toList());

        entity.setFuelTanks(fuelTankEntities);

        return entity;
    }
}